plugin/pkg/kmsproviders/vault/client.go

```diff
@@ -0,0 +1,212 @@
+	encryptPath string
+	decryptPath string
+	authPath    string
+}
+
+// Initialize a client wrapper for vault kms provider.
+func newClientWrapper(config *EnvelopeConfig) (*clientWrapper, error) {
+	client, err := newVaultClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Vault transit path is configurable. "path", "/path", "path/" and "/path/"
+	// are the same.
+	transit := defaultTransitPath
+	if config.TransitPath != "" {
+		transit = config.TransitPath
+	}
+
+	// auth path is configurable. "path", "/path", "path/" and "/path/" are the same.
+	auth := defaultAuthPath
+	if config.AuthPath != "" {
+		auth = config.AuthPath
+	}
+	wrapper := &clientWrapper{
+		client:      client,
+		encryptPath: path.Join("v1", transit, "encrypt"),
+		decryptPath: path.Join("v1", transit, "decrypt"),
+		authPath:    path.Join(auth),
+	}
+
+	// Set token for the vaultapi.client.
+	if len(config.Token) != 0 {
+		client.SetToken(config.Token)
+	} else {
+		err = wrapper.refreshToken(config)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapper, nil
+}
+
+func newVaultClient(config *EnvelopeConfig) (*vaultapi.Client, error) {
+	vaultConfig := vaultapi.DefaultConfig()
+	vaultConfig.Address = config.Address
+
+	tlsConfig := &vaultapi.TLSConfig{
+		CACert:        config.VaultCACert,
+		ClientCert:    config.ClientCert,
+		ClientKey:     config.ClientKey,
+		TLSServerName: config.TLSServerName,
+	}
+	if err := vaultConfig.ConfigureTLS(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return vaultapi.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to vaultapi.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig) error {
+	switch {
+	case config.ClientCert != "" && config.ClientKey != "":
+		token, err := c.tlsToken(config)
+		if err != nil {
+			return fmt.Errorf("rotating token through TLS auth backend: %v", err)
+		}
+		c.client.SetToken(token)
+	case config.RoleID != "":
+		token, err := c.appRoleToken(config)
+		if err != nil {
+			return fmt.Errorf("rotating token through app role backend: %v", err)
+		}
+		c.client.SetToken(token)
+	default:
+		// configuration has already been validated, flow should not reach here
+		return errors.New("the Vault authentication configuration is invalid")
+	}
+
+	return nil
+}
+
+func (c *clientWrapper) tlsToken(config *EnvelopeConfig) (string, error) {
+	resp, err := c.client.Logical().Write("/"+path.Join(c.authPath, "cert", "login"), nil)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) appRoleToken(config *EnvelopeConfig) (string, error) {
+	data := map[string]interface{}{
+		"role_id":   config.RoleID,
+		"secret_id": config.SecretID,
+	}
+	resp, err := c.client.Logical().Write("/"+path.Join(c.authPath, "approle", "login"), data)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) decrypt(keyName string, cipher string) (string, error) {
```

Comment:
`decrypt` and `encrypted` have to be called with the read lock aquired, correct?  If so `decryptLocked`,

Reply:
done